File: Dave Min/2023-05-07-BinarySearch.cs
public class Solution {
    public int Search(int[] nums, int target) {        
        //1. Recursive Binary Search
        return BSearch(nums, target, 0, nums.Length-1);

        //2. Built-in Binary Search
        //int result = nums.ToList().BinarySearch(target);
        //return  result>=0?result:-1;
    }
    public int BSearch(int[] nums, int target, int x, int y) {
        if(x>y) return -1;
        int mid = (y+x)/2;

        if(nums[mid] == target) return mid;
        else if(nums[mid] < target) return BSearch(nums, target, mid+1, y);
        else return BSearch(nums, target, x, mid-1);
    }
}


//1. Recursive Binary Search 
//Time complexity: O(log n);
//Space complexity: O(log n);

//2. Built-in Binary Search
//Time complexity: O(log n);
//Space complexity: O(1);

File: Ellen Jung/2023-05-07-binary-search.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        index = -1
        first = 0
        end = len(nums) - 1
        while first <= end:
            mid = first + (end - first)/2
            if nums[mid] == target:
                index = mid
                break;
            elif nums[mid] < target:
                first = mid + 1
            elif nums[mid] > target:
                end = mid - 1

        return index